### PR TITLE
[dist, trainer, ci] feat: support Qwen3-VL FSDP2 torch compile

### PR DESCRIPTION
**Pull request by WenzheWang**
### What does this PR do?

Enable opt-in per-block `torch.compile` for dense Qwen3-VL training on CUDA with FSDP2. The implementation compiles only `Qwen3VLTextDecoderLayer.forward` before sharding; the vision tower, DeepStack injection, and language-model head stay eager.

Unsupported combinations fail explicitly rather than silently compiling a partial or unstable path. Qwen3-VL CUDA Graph replay, Qwen3-VL-MoE, ChunkMBS, Ulysses sequence parallelism, context parallelism, asynchronous tensor parallelism, ExtraParallel, DDP/non-FSDP execution, dynamic compile shapes, NPU execution, and other multimodal model families remain unsupported.

### Checklist Before Starting

- Searched open VeOmni PRs and the exact owner-qualified head branch; no duplicate or overlapping Qwen3-VL compile implementation was found.
- #933 explicitly leaves `torch.compile` unsupported for its Qwen3-VL ChunkMBS/SP/EP scope. This PR rejects ChunkMBS and ExtraParallel, so it is independent of #933.
- PR title follows the enforced `[{modules}] {type}: {description}` format.

### Test

- `python -m pytest -q tests/distributed/test_torch_compile.py`: **43 passed, 1 skipped**. The skipped CUDA case exercises Inductor/fullgraph parity for packed FlashAttention input and is selected by the GPU unit-test workflow.
- `python -m pytest -q tests/distributed`: **48 passed, 11 skipped**.
- `python -m pytest -q tests/trainer/test_step_sync.py`: **2 passed**.
- `python -m pytest -q tests/models/test_vlm_trainer.py`: **12 passed**.
- `ruff check tasks tests veomni docs`: passed; **622 files** checked.
- `pre-commit run check-yaml --files .github/workflows/gpu_unit_tests.yml`: passed.
- `python tests/special_sanity/check_device_api_usage.py -d veomni`: passed.
- `python tests/special_sanity/check_device_api_usage.py -d tests`: passed.
- `python scripts/ci/check_doc_task_paths.py`: passed.
- `git diff --check`: passed.
- Independent review: **safe**, with no remaining findings.

#### Two-A800 full training-step validation

Validated commit `bdb20ff7` on 2 x NVIDIA A800 80GB PCIe with FSDP2/DP-shard size 2, BF16 parameters, FP32 reductions, non-reentrant gradient checkpointing, and AdamW. Each run exercised the complete multimodal forward, loss, backward, and optimizer path on deterministic synthetic text plus image inputs using the repository's reduced Qwen3-VL config (2 text decoder layers, 2 vision blocks). Two symmetric seed pairs each ran 2 warmup and 5 measured optimizer steps.

| Seed | Max paired loss rel. diff | Exact final-parameter delta L2 / reference L2 | Steady-state speedup | Peak allocated delta |
| --- | ---: | ---: | ---: | ---: |
| 20260804 | 0.02157% | 0.005238% | 1.0441x | 0 GiB |
| 20260805 | 0.02868% | 0.005185% | 1.0347x | 0 GiB |

All losses and gradients were finite. Both compile runs captured one Dynamo graph with 83 calls, compiled exactly the two text decoder layers, and compiled zero vision blocks. The mean exploratory steady-state speedup was **1.0394x** (**3.94% higher throughput**, equivalent to **3.79% lower step time**) with identical 5.987 GiB peak allocated memory.

The A800 node's driver 535 cannot run the repository's current PyTorch 2.11/cu130 lock stack, so this hardware validation reused a compatible PyTorch 2.6.0+cu124 environment with Transformers 5.9.0. These results provide reduced semantic and exploratory performance evidence for commit `bdb20ff7`; they do not claim Qwen3-VL-8B capacity, convergence, validation of the current cu130 lock stack, or execution of the later CUDA regression test added in commit `330be524`.

### API and Usage Example

```shell
bash train.sh tasks/train_vlm.py configs/multimodal/qwen3_vl/qwen3_vl_dense.yaml \
    --model.model_path ./Qwen3-VL-8B-Instruct \
    --data.train_path ./sharegpt4v_instruct_gpt4-vision_cap100k_coco.json \
    --train.dyn_bsz true \
    --train.pad_to_length true \
    --train.torch_compile.enable true
```

The default remains disabled. Qwen3-VL currently requires CUDA, FSDP2, fixed padded packed inputs, the default `train.torch_compile.backend=inductor` and `train.torch_compile.mode=None`, `train.torch_compile.dynamic=false`, `train.accelerator.ulysses_size=1`, `train.accelerator.cp_size=1`, and `train.accelerator.enable_async=false`. Packed FlashAttention boundary metadata may create separate Inductor specializations, so CUDA Graph replay is rejected for this model path.

### Design & Code Changes

- Detect multimodal models and allow-list only dense `qwen3_vl` with `Qwen3VLTextDecoderLayer` ownership.
- Validate the model family immediately after model construction, before freeze/data setup, so unsupported configurations fail early.
- Validate CUDA, FSDP2, sequence/context/async parallelism, ExtraParallel, and CUDA Graph Trees compatibility before compilation.
- Compile decoder forwards before FSDP2 sharding while leaving cross-modal components eager.
- Mark VLM micro-steps for CUDA Graph Trees when the selected compile mode uses them.
- Enable the VLM static padded-data contract and retain explicit argument validation.
- Add CPU fullgraph traceability, fail-closed runtime-contract, and CUDA Inductor packed-FlashAttention parity coverage.
- Run the focused compile suite in the GPU unit-test workflow and update Qwen3-VL usage documentation.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- [x] Applied pre-commit checks
- [x] Added/updated documentation
- [x] No `tasks/` training scripts were moved or renamed
- [x] Added focused CPU and GPU coverage to the distributed torch-compile test suite

### CODE REVIEW

File: veomni/distributed/torch_compile.py [hunk 1]
```diff
@@ -49,6 +54,57 @@ def _decoder_block_class_names(model: nn.Module) -> set[str]:
...
+    model_type = getattr(config, "model_type", None)
+    expected_decoder_block = _SUPPORTED_MULTIMODAL_DECODER_BLOCKS.get(model_type)
+    if expected_decoder_block is None:
+        raise RuntimeError(
+            "train.torch_compile.enable currently supports multimodal training only for dense Qwen3-VL "
+            f"(model_type='qwen3_vl'); got model_type={model_type!r}."
+        )
+    if compile_config.dynamic:
+        raise RuntimeError("train.torch_compile.enable for Qwen3-VL requires train.torch_compile.dynamic=False.")
+    if sequence_parallel_enabled:
```
[the remark this answers is not in the document]
Comment (WenzheWang) [reply veomni/distributed/torch_compile.py]: Fixed in 330be524. The compile guard now receives both sequence-parallel and async state, rejects Ulysses, context parallelism, and async tensor parallelism explicitly, and names all three configuration knobs in the error. Parameterized regression coverage exercises each mode.

File: veomni/distributed/torch_parallelize.py [hunk 2]
```diff
@@ -249,7 +249,11 @@ def parallelize_model_fsdp2(
             )
         validate_compile_config_for_fsdp2(compile_config, enable_reshard_after_forward)
 
-        compiled_count = compile_decoder_blocks(model, compile_config)
+        compiled_count = compile_decoder_blocks(
+            model,
+            compile_config,
+            sequence_parallel_enabled=parallel_state.sp_enabled,
```
[the remark this answers is not in the document]
Comment (WenzheWang) [reply veomni/distributed/torch_parallelize.py:287]: Fixed in 330be524. `torch_parallelize` now passes `parallel_state.async_enabled` alongside `parallel_state.sp_enabled`, keeping model validation free of a global state lookup.

File: tests/distributed/test_torch_compile.py [hunk 3]
```diff
@@ -120,6 +136,118 @@ def fake_compile(fn, **kwargs):
...
+    hidden_states_compiled = hidden_states_eager.detach().clone().requires_grad_(True)
+    position_ids = torch.arange(7, dtype=torch.long).view(1, 7)
+    rotary_emb = Qwen3VLTextRotaryEmbedding(config)
+    position_embeddings = rotary_emb(hidden_states_eager, position_ids)
+
+    with use_parallel_state(SimpleNamespace(async_enabled=False)):
+        assert (
+            compile_decoder_blocks(
+                compiled_model,
+                CompileConfig(enable=True, backend="eager", fullgraph=True, dynamic=False),
```
[the remark this answers is not in the document]
Comment (WenzheWang) [reply tests/distributed/test_torch_compile.py:261]: Fixed in 330be524. The CPU case is now named `test_qwen3_vl_decoder_traces_under_fullgraph`. A CUDA-only Inductor/fullgraph test covers packed FlashAttention position IDs and compares eager versus compiled outputs, input gradients, and parameter gradients. The focused test is also selected by the GPU unit-test workflow.

File: tests/distributed/test_torch_compile.py [hunk 4]
```diff
@@ -120,6 +136,118 @@ def fake_compile(fn, **kwargs):
...
+            layer._gradient_checkpointing_func = partial(checkpoint, use_reentrant=False)
+
+    compiled_model = ToyQwen3VLModel(compiled_layer)
+    hidden_states_eager = torch.randn(1, 7, config.hidden_size, requires_grad=True)
+    hidden_states_compiled = hidden_states_eager.detach().clone().requires_grad_(True)
+    position_ids = torch.arange(7, dtype=torch.long).view(1, 7)
+    rotary_emb = Qwen3VLTextRotaryEmbedding(config)
+    position_embeddings = rotary_emb(hidden_states_eager, position_ids)
+
+    with use_parallel_state(SimpleNamespace(async_enabled=False)):
```
[the remark this answers is not in the document]
Comment (WenzheWang) [reply tests/distributed/test_torch_compile.py:257]: Fixed in 330be524. Added a load-bearing comment explaining that an installed parallel state prevents the warning/logging branch from entering the fullgraph trace.

File: veomni/arguments/arguments_types.py [hunk 5]
```diff
@@ -1500,12 +1500,12 @@ def __post_init__(self):
                 )
             if not getattr(self.data, "supports_torch_compile", True):
                 raise ValueError(
-                    "train.torch_compile.enable currently supports text trainers only. "
-                    "Multimodal/DiT/Omni data pipelines do not implement pad_to_length for static packed shapes yet."
+                    "train.torch_compile.enable is not supported by this data pipeline. "
```
[the remark this answers is not in the document]
Comment (WenzheWang) [reply veomni/arguments/arguments_types.py:1545]: Fixed in 330be524. `VLMTrainer` now validates compile model and runtime compatibility immediately after `_build_model()`, before freeze handling, asset processing, and dataset construction. A regression test verifies unsupported multimodal families fail before asset initialization.

**Comment by WenzheWang** [comment]
Updated through `d40dfc9d` and refreshed onto `main@7bb4b621`:

- reject context parallelism and asynchronous tensor parallelism in addition to Ulysses;
- thread `async_enabled` explicitly through compile validation;
- validate unsupported VLM families immediately after model construction;
- add CUDA Inductor/fullgraph parity coverage for packed FlashAttention inputs;
- run the focused compile suite in the GPU unit-test workflow; and
- replace the direct device-type literal with VeOmni's unified device API, with a CUDA-unavailable regression test;
- preserve the current main step-sync and global loss-token lifecycle in `VLMTrainer`; and
- reject Qwen3-VL CUDA Graph replay explicitly because packed FlashAttention metadata can vary between micro-batches.

Validation on the current tree:

- focused compile suite: `43 passed, 1 skipped`;
- all distributed tests: `48 passed, 11 skipped`;
- main step-sync tests: `2 passed`;
- VLM trainer tests: `12 passed`;
- device-API sanity for both `veomni` and `tests`: passed;
- docs task-path check, Ruff quality, and `git diff --check`: passed;
- independent review: `safe`.

The CUDA Inductor packed-FlashAttention case is selected by the GPU workflow and was skipped in the local CPU-only run. All five review threads remain answered and resolved.

Performance evidence remains anchored to the reduced two-A800 experiment on `bdb20ff7`: the two symmetric repetitions measured `1.0441x` and `1.0347x` steady-state speedups. Their mean is **1.0394x** (**3.94% higher throughput**, equivalent to **3.79% lower step time**) with unchanged 5.987 GiB peak allocated memory. This is exploratory reduced-model evidence, not a Qwen3-VL-8B capacity, convergence, or current cu130-stack result.

**Review by Coach257 — APPROVED** [review]

**Comment by WenzheWang** [comment]
@Coach257 The branch has been updated to the latest `main`, and the remaining review feedback has been addressed in `d40dfc9d`. The PR is currently mergeable, with local validation and the non-hardware checks passing; the GPU and NPU jobs are still queued.

Could you please review the latest changes and merge the PR once the remaining CI checks pass? Thanks.
